opentelemetry-sdk/src/opentelemetry/sdk/metrics/_internal/export/__init__.py

```diff
@@ -411,22 +411,17 @@ def shutdown(self, timeout_millis: float = 30_000, **kwargs) -> None:
         pass
 
 
-class PeriodicExportingMetricReader(MetricReader):
-    """`PeriodicExportingMetricReader` is an implementation of `MetricReader`
-    that collects metrics based on a user-configurable time interval, and passes the
-    metrics to the configured exporter.
-
-    The configured exporter's :py:meth:`~MetricExporter.export` method will not be called
-    concurrently.
+class ExportingMetricReader(MetricReader):
```

Comment:
I believe this is definitely a valid use case. Would it make more sense to add this functionality as configuration to the `PeriodicExportingMetricReader` instead (e.g., set time interval to 0 to not create a worker thread) and ask the user to call `force_flush()` explicitly to replicate this behavior?

One big reason to do the above would be to not introduce any new API elements that we must support and that could potentially have breaking changes.

Reply:
> I believe this is definitely a valid use case. Would it make more sense to add this functionality as configuration to the `PeriodicExportingMetricReader` instead (e.g., set time interval to 0 to not create a worker thread) and ask the user to call `force_flush()` explicitly to replicate this behavior?
> 
> One big reason to do the above would be to not introduce any new API elements that we must support and that could potentially have breaking changes.

I'd say if PeriodicExportingMetricReader would support 0 interval as 'no interval' state, and that user can use force_flush() to push the collection and metric export explicitly, it sure would fit the bill!

However, for aesthetic(!) reasons, using `Periodic`ExportingMetricReader with zero interval would sort of feel kind of 'irregular' to how it is supposed to be used, and without proper understanding, someone looking at the code without knowledge may be confused by looking at its zero interval and use of `force_flush()` all the time.

Also, I did not introduce anything different or breaking with this PR, but merely relocated existing logic in a more natural way, that ExportingMetricReader would have the non-periodic elements, while PeriodicExportingMetricReader would inherit all of it plus adding the periodic triggering nature to it - which is actually a more appropriate object-oriented way of designing and implementing it, in my opinion.

I've created the pytest for ExportingMetricReader, which seemed to pass the test, as well as the existing tests for PeriodicExportingMetricReader also passed without any problems, so although I understand your concern for the change, this change is actually backward compatible, and have very little chance of breaking anything existing, in my opinion. I do really wish the people could use ExportingMetricReader for non-periodic collection/export, and choose PeriodicExportingMetricReader for periodic reporting purposes. It does make more sense in that way.

Reply:
For breaking changes, I meant more the fact that `ExportingMetricReader ` is not explicitly spec'ed out in the OpenTelemetry specifications. This means that we are exposing a new api in which we would have to support that is solely in the Python implementation (which would require additional maintenance and support from the community/maintainers). This might not be ideal, especially because the same functionality can already be achieved with existing APIs, just maybe documentation could be added to outline this particular use case.

Reply:
@aabmass @srikanthccv  @ocelotl thoughts?

Reply:
>The [MetricReader.Collect](https://github.com/open-telemetry/opentelemetry-python/pull/3059/files#collect) method allows general-purpose MetricExporter instances to explicitly initiate collection, commonly used with pull-based metrics collection. A common sub-class of MetricReader, the periodic exporting MetricReader SHOULD be provided to be used typically with push-based metrics collection.

Re-reading the spec again, it seems to me that our `MetricReader` should itself be sufficient for Howard's use-case, but it's not today directly useable since there are some abstract methods. I think this is a spec non-compliant? and if we fix that, it will address both problems.

Reply:
> > The [MetricReader.Collect](https://github.com/open-telemetry/opentelemetry-python/pull/3059/files#collect) method allows general-purpose MetricExporter instances to explicitly initiate collection, commonly used with pull-based metrics collection. A common sub-class of MetricReader, the periodic exporting MetricReader SHOULD be provided to be used typically with push-based metrics collection.
> 
> Re-reading the spec again, it seems to me that our `MetricReader` should itself be sufficient for Howard's use-case, but it's not today directly useable since there are some abstract methods. I think this is a spec non-compliant? and if we fix that, it will address both problems.

Hi @srikanthccv , you are right! The only problem that I have with `MetricReader` is that it has those abstract methods that prevent me from using it directly. If it is possible to make those methods non-abstract, and if that is something that wouldn't make anyone unhappy, that would be great.